### PR TITLE
Merge wiring

### PR DESCRIPTION
**Pull request by Jacobsolawetz**
```
%%writefile -a newmerge.yaml
slices:
  - sources:
      - model: psmathur/orca_mini_v3_13b
        layer_range: [0, 40]
      - model: garage-bAInd/Platypus2-13B
        layer_range: [0, 40]
merge_method: slerp
base_model: psmathur/orca_mini_v3_13b
parameters:
  t:
    - filter: self_attn
      value: [0, 0.5, 0.3, 0.7, 1]
    - filter: mlp
      value: [1, 0.5, 0.7, 0.3, 0]
    - value: 0.5 # fallback for rest of tensors
dtype: float16
```

```
arcee.mergekit_yaml("new_merge3", "newmerge.yaml")
```

```
arcee.mergekit_evolve(
    merging_name="dummy_merging16",
    arcee_pretrained_models=["sliced-slack-2"],
    arcee_aligned_models=[],
    arcee_merged_models=[],
    hf_models=["arcee-ai/mistral-sliced"],
    arcee_eval_qa_set_names_and_weights=[{"tax_questions": 0.5}, {"tax_questions": 1}],
    general_evals_and_weights=[{"agieval_lsat_ar": 1}],
    base_model="arcee-ai/mistral-sliced",
    merge_method="ties",
    time_budget_secs=864000,
    wandb_key="cacf018fbe82e6e1a6f3174064f1c71a810aae5c",
    target_compute="p5.48xlarge",
    capacity_id="cr-0190d9df55a87083b"
)
```

### CODE REVIEW

File: arcee/api.py [hunk 1]
```diff
@@ -82,26 +125,85 @@ def start_pretraining(pretraining_name: str, corpus: str, base_model: str) -> No
...
+    capacity_id: str = None,
+    time_budget_secs: int = 1,
+) -> None:
     """
     Start merging models
 
     Args:
         merging_name (str): The name of the merging job
-        arcee_models (list): A list of ARCEE models to merge
+        arcee_aligned_models (list): A list of ARCEE models to merge
```
Comment (tleyden) [inline arcee/api.py:178]: I wouldn't capitalize "ARCEE" here - I'd suggest either "Arcee" or "Arcee.ai".

Does this the model ids or the model names?  Are the model names unique?  (I think they are, just checking)

What happens if this is empty?

File: arcee/api.py [hunk 2]
```diff
@@ -82,26 +125,85 @@ def start_pretraining(pretraining_name: str, corpus: str, base_model: str) -> No
...
+        arcee_merged_models (list): A list of ARCEE models already merged
+        arcee_pretrained_models (list): A list of pretrained ARCEE models
         hf_models (list): A list of Hugging Face models to merge
-        qa_set_ids (list): A list of QA sets to merge
-        general_evals (list): A list of general evaluations to merge
+        eval_qa_set_names_and_weights (list): A list of QA set names to merge
```
Comment (tleyden) [inline arcee/api.py:182]: This doesn't work yet right?  We should probably add a disclaimer to that effect

File: arcee/api.py [hunk 3]
```diff
@@ -82,26 +125,85 @@ def start_pretraining(pretraining_name: str, corpus: str, base_model: str) -> No
...
+        general_evals_and_weights (list): A list of general evaluations to merge
```
Comment (tleyden) [inline arcee/api.py:183]: As a casual user, how would I get the list of general evaluations?  Are there any selected by default if nothing is passed here?  (that might be a nicer UX)

File: arcee/api.py [hunk 4]
```diff
@@ -82,26 +125,85 @@ def start_pretraining(pretraining_name: str, corpus: str, base_model: str) -> No
...
         base_model (str): The name of the base model to use
         merge_method (str): The merging method to use
```
Comment (tleyden) [inline arcee/api.py:185]: Can we point to the list of available merging methods?  Also a sensible default might make it easier to use.

File: arcee/api.py [hunk 5]
```diff
@@ -82,26 +125,85 @@ def start_pretraining(pretraining_name: str, corpus: str, base_model: str) -> No
...
+    merging_name: str,
+    wandb_key: str = None,
+    arcee_aligned_models: Optional[List[str]] = None,
+    arcee_merged_models: Optional[List[str]] = None,
+    arcee_pretrained_models: Optional[List[str]] = None,
+    hf_models: Optional[List[str]] = None,
+    arcee_eval_qa_set_names_and_weights: Optional[List[dict]] = None,
+    general_evals_and_weights: Optional[List[dict]] = None,
+    base_model: Optional[str] = None,
+    merge_method: Optional[str] = None,
```
Comment (tleyden) [inline arcee/api.py:168]: I think a default `merge_method` here would make it easier to use, like `["dare_ties"]`

In general anywhere we can add default vals increases the "Don't make me think" UX

File: arcee/api.py [hunk 6]
```diff
@@ -82,26 +125,85 @@ def start_pretraining(pretraining_name: str, corpus: str, base_model: str) -> No
...
+        arcee_pretrained_models (list): A list of pretrained ARCEE models
         hf_models (list): A list of Hugging Face models to merge
-        qa_set_ids (list): A list of QA sets to merge
-        general_evals (list): A list of general evaluations to merge
+        eval_qa_set_names_and_weights (list): A list of QA set names to merge
+        general_evals_and_weights (list): A list of general evaluations to merge
         base_model (str): The name of the base model to use
         merge_method (str): The merging method to use
-        time_budget (int): The time budget for the merging job (hourss)
+        time_budget_secs (int): The time budget for the merging job (seconds)
```
Comment (tleyden) [inline arcee/api.py:186]: What happens when the budget is exceeded?

File: arcee/api.py [hunk 7]
```diff
@@ -33,17 +37,55 @@ def upload_qa_pairs(qa_set: str, qa_pairs: List[Dict[str, str]]) -> Dict[str, st
...
+def chunk_list(lst, chunk_size):
+    """Yield successive n-sized chunks from lst."""
+    for i in range(0, len(lst), chunk_size):
+        yield lst[i:i + chunk_size]
+
+def upload_instructions_from_csv(qa_set: str, csv_path: str, prompt_column: str = "prompt", completion_column: str = "completion", batch_size: int = 200) -> None:
+    """
+    Upload QA pairs from a CSV file to a specific QA set.
+
+    Args:
```
Comment (tleyden) [inline arcee/api.py:61]: Can you add docs for the `prompt_column` and `completion_column`?

File: arcee/api.py [hunk 8]
```diff
@@ -33,17 +37,55 @@ def upload_qa_pairs(qa_set: str, qa_pairs: List[Dict[str, str]]) -> Dict[str, st
...
+        Dict[str, str]: The response from the make_request call.
+    """
+    qa_pairs = []
+
+    print(f"Reading QA pairs from {csv_path}...")
+    with open(csv_path, 'r', encoding='utf-8') as csvfile:
+        reader = csv.DictReader(csvfile)
+        for row in reader:
+            if prompt_column not in row.keys() or completion_column not in row.keys():
+                #raise Exception(f"Each row must have a '{question_column}' and an '{answer_column}' key. You can override the column names using the question_column and answer_column arguments.")
```
Comment (tleyden) [inline arcee/api.py:75]: Any reason to leave this commented code in?

File: arcee/api.py (hunk 7 above)
Comment (tleyden) [reply arcee/api.py:61]: It would be good to tell the user that these column names will be propagated to the system.

File: arcee/api.py [hunk 9]
```diff
@@ -33,17 +37,55 @@ def upload_qa_pairs(qa_set: str, qa_pairs: List[Dict[str, str]]) -> Dict[str, st
...
+                raise Exception(f"Each row must have a '{prompt_column}' and an '{completion_column}' key. You can override the column names using the prompt_column and completion_column arguments.")
+            qa_pairs.append({
+                f"{prompt_column}": row[prompt_column],
+                f"{completion_column}": row[completion_column]
```
Comment (tleyden) [inline arcee/api.py:79]: Will this even work?  I thought we expected 

```
{
  "question": "..",
  "answer": ".."
}
```

Will it make it into the db?  

Any chance you could upload a CSV or a few CSVs you tested with to the PR?  Or add more details on how you tested the code

File: arcee/api.py [hunk 10]
```diff
@@ -82,26 +125,85 @@ def start_pretraining(pretraining_name: str, corpus: str, base_model: str) -> No
...
+def mergekit_yaml(
+    merging_name: str,
+    merging_yaml_path: str
+) -> None:
+    """
+    Start merging models
+
+    Args:
+        merging_name (str): The name of the merging job
+        merging_yaml (str): The yaml file containing the merging instructions
```
Comment (tleyden) [inline arcee/api.py:139]: Can you link to the yaml spec or paste an example inline in the docs?

File: arcee/api.py [hunk 11]
```diff
@@ -82,26 +125,85 @@ def start_pretraining(pretraining_name: str, corpus: str, base_model: str) -> No
...
+    arcee_merged_models: Optional[List[str]] = None,
+    arcee_pretrained_models: Optional[List[str]] = None,
+    hf_models: Optional[List[str]] = None,
+    arcee_eval_qa_set_names_and_weights: Optional[List[dict]] = None,
+    general_evals_and_weights: Optional[List[dict]] = None,
+    base_model: Optional[str] = None,
+    merge_method: Optional[str] = None,
+    target_compute: str = None,
+    capacity_id: str = None,
+    time_budget_secs: int = 1,
```
Comment (tleyden) [inline arcee/api.py:171]: 1 second time budge by default?  Shouldn't this be at least an hour?

File: arcee/api.py [hunk 12]
```diff
@@ -82,26 +125,85 @@ def start_pretraining(pretraining_name: str, corpus: str, base_model: str) -> No
...
+        data = {
+            "merging_name": merging_name,
+            "best_merge_yaml": str(merging_yaml)
+        }
+
+        return make_request("post", Route.merging + "/startMerging", data)
+
+def mergekit_evolve(
+    merging_name: str,
+    wandb_key: str = None,
```
Comment (tleyden) [inline arcee/api.py:160]: Can you add this to the documented args.  Also `Optional[str] = None` might be better to match the other args

File: arcee/api.py [hunk 13]
```diff
@@ -33,17 +37,55 @@ def upload_qa_pairs(qa_set: str, qa_pairs: List[Dict[str, str]]) -> Dict[str, st
         Dict[str, str]: The response from the make_request call.
     """
     if len(qa_pairs) > 2000:
-        raise Exception("You can only upload 2000 QA pairs at a time")  
+        raise Exception("You can only upload 2000 QA pairs at a time")
```
Comment (tleyden) [inline arcee/api.py:40]: Should we remove this limitation?

File: arcee/api.py [hunk 14]
```diff
@@ -33,17 +37,55 @@ def upload_qa_pairs(qa_set: str, qa_pairs: List[Dict[str, str]]) -> Dict[str, st
...
+                f"{prompt_column}": row[prompt_column],
+                f"{completion_column}": row[completion_column]
+            })
+
+    print(f"Total QA pairs read: {len(qa_pairs)}")
+    # Split the QA pairs into chunks and upload each chunk separately
+    for i, chunk in enumerate(chunk_list(qa_pairs, batch_size)):
+        print(f"Uploading chunk {i + 1} of {len(qa_pairs) // batch_size + 1}...")
+        
+        upload_qa_pairs(qa_set=qa_set, qa_pairs=chunk, prompt_column=prompt_column, completion_column=completion_column)
```
Comment (tleyden) [inline arcee/api.py:87]: Style nit, I think this is more readable:

```
upload_qa_pairs(
  qa_set=qa_set, 
  qa_pairs=chunk,
  ..
)
```

**Review by tleyden — CHANGES_REQUESTED** [review]
Looks good overall.   Any chance we can remove that 2K pairs limitation?  

How about on our public facing api we have no limitation, and that calls an internal API with a chunking method?

eg:

```
def upload_qa_pairs(qa_set: str, qa_pairs: List[Dict[str, str]], prompt_column: str = "prompt", completion_column: str = "completion") -> Dict[str, str]:
```

would be the public api you could call with 100k pairs.  It will chunk into 2k chunks and call:

```
def _internal_upload_qa_pairs(qa_set: str, qa_pairs: List[Dict[str, str]], prompt_column: str = "prompt", completion_column: str = "completion") -> Dict[str, str]:
```

which has a 2K limit.  All of our code would call `_internal_upload_qa_pairs` and handle the chunking.

Otherwise it's confusing that `upload_qa_pairs` has a 2K limit, but `upload_instructions_from_csv` does not.

File: arcee/api.py [hunk 15]
```diff
@@ -33,17 +37,55 @@ def upload_qa_pairs(qa_set: str, qa_pairs: List[Dict[str, str]]) -> Dict[str, st
...
     data = {"qa_set_name": qa_set, "qa_pairs": qa_list}
-    return make_request("post", Route.alignment+"/qaUpload", data)
+    return make_request("post", Route.alignment + "/qaUpload", data)
+
+def chunk_list(lst, chunk_size):
+    """Yield successive n-sized chunks from lst."""
+    for i in range(0, len(lst), chunk_size):
+        yield lst[i:i + chunk_size]
+
+def upload_instructions_from_csv(qa_set: str, csv_path: str, prompt_column: str = "prompt", completion_column: str = "completion", batch_size: int = 200) -> None:
```
Comment (tleyden) [inline arcee/api.py:57]: This is confusing - why is it called `qa_pairs` in other methods, and `instructions` here?  I would say its better to stay consistent and call it `upload_qa_pairs_from_csv` for now.

Then later we could change it to `instructions` everywhere if we want a more general term

File: arcee/api.py (hunk 2 above)
Comment (Jacobsolawetz) [reply arcee/api.py:182]: Oh @tleyden it works my man!

File: arcee/api.py (hunk 3 above)
Comment (Jacobsolawetz) [reply arcee/api.py:183]: Ok yep, I agree... I'll put a default in there

**Comment by Jacobsolawetz** [comment]
> would be the public api you could call with 100k pairs. It will chunk into 2k chunks and call:

@tleyden this sounds good but going to suggest out of scope for this PR

File: arcee/api.py (hunk 2 above)
Comment (Jacobsolawetz) [reply arcee/api.py:182]: Present in training PR

File: arcee/api.py (hunk 1 above)
Comment (roninCode) [reply arcee/api.py:178]: Model names, afaik are unique per their table (idk if you can name a pretraining and an align the same thing though...)

What does clean mean here? Like convert all spaces to `-`, no special chars?
Does this api support the old naming conventions we have been using for models and should it?
Where do you see this restriction taking place - api/platform ui?

File: arcee/api.py (hunk 15 above)
Comment (Jacobsolawetz) [reply arcee/api.py:57]: Good call

Comment (Jacobsolawetz) [reply arcee/api.py:57]: This is the wrong way to do it, should be one unified flip

Comment (Jacobsolawetz) [reply arcee/api.py:57]: cc @mmcquade11

File: arcee/api.py (hunk 9 above)
Comment (Jacobsolawetz) [reply arcee/api.py:79]: Due to the remap in upload_qa_pairs, it still does

File: arcee/api.py (hunk 1 above)
Comment (Jacobsolawetz) [reply arcee/api.py:178]: UI editor + API

**Pull request merged**